frappe/integrations/doctype/token_cache/token_cache.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2019, Frappe Technologies and contributors
+# For license information, please see license.txt
+
+from __future__ import unicode_literals
+import frappe
+import requests
+from urllib.parse import urlencode
+from datetime import datetime, timedelta
+from frappe.model.document import Document
+
+class TokenCache(Document):
+
+	def get_auth_header(self):
+		if self.access_token:
+			headers = {'Authorization': 'Bearer ' + self.access_token}
+			return headers
+
+		raise frappe.exceptions.DoesNotExistError
+
+	def check_validity(self):
+		if(self.get('__islocal') or (not self.access_token)):
+			raise frappe.exceptions.DoesNotExistError
+
+		if not self.is_expired():
+			return self
+
+		return self.refresh_token()
+
+	def refresh_token(self):
+		app = frappe.get_doc("Connected App", self.connected_app)
+		oauth = app.get_oauth2_session()
+		new_token = oauth.refresh_token(
+			app.token_endpoint,
+			client_secret=app.get_password('client_secret'),
+			token=self.get_json()
+		)
+
+		if new_token.get('access_token') and app.revocation_endpoint:
+			# Revoke old token
+			requests.post(
+				app.revocation_endpoint,
+				data=urlencode({'token': new_token.get('access_token')}),
+				headers={
+					'Authorization': 'Bearer ' + new_token.get('access_token'),
```

Comment:
revoke endpoint headers are not standardized.

Some providers use no headers
Some providers use new valid token
Some providers use Basic Auth with client_id as username and client_secret as password.